File: nanozero-training/src/nanozero_training/network/move_encoding.py
"""Move encoding AlphaZero (Python) — parité bit-perfect vs nanozero-nn MoveEncoding Java.

Reproduit BIT-POUR-BIT `nanozero-nn/src/main/java/.../MoveEncoding.java::encode`
(cf. SPEC §3.5, §5.5 nn). Conformité ADR-010 training (duplication contrôlée
des conventions Java côté Python self-play).

Format figé : 73 plans par from-square x 64 from-squares = 4672 indices.

- Plans 0..55 : queen-style (8 directions x 7 distances)
- Plans 56..63 : knight (8 deltas)
- Plans 64..72 : underpromotions (3 directions x 3 sous-pièces N/B/R)

Index final : `from * POLICY_PLANES + plane_index`. Mirror XOR 56 si
`board.turn == BLACK` (perspective P1 normalisée — le réseau prédit toujours
du point de vue du côté au trait).
"""

from __future__ import annotations

import chess
import numpy as np
import numpy.typing as npt

# Constants (figées, alignées MoveEncoding.java).
N_FROM_SQUARES = 64
N_QUEEN_PLANES = 56  # 8 directions x 7 distances
N_KNIGHT_PLANES = 8
N_UNDERPROMOTION_PLANES = 9  # 3 directions x 3 pieces (N/B/R)
POLICY_PLANES = N_QUEEN_PLANES + N_KNIGHT_PLANES + N_UNDERPROMOTION_PLANES  # 73
POLICY_INDICES = POLICY_PLANES * N_FROM_SQUARES  # 4672

# Queen-style directions (deltaRank, deltaFile), ordre NORMATIF aligné Java
# `MoveEncoding.QUEEN_DELTAS` : N, NE, E, SE, S, SW, W, NW.
QUEEN_DELTAS: tuple[tuple[int, int], ...] = (
    (+1, 0),
    (+1, +1),
    (0, +1),
    (-1, +1),
    (-1, 0),
    (-1, -1),
    (0, -1),
    (+1, -1),
)

# Knight deltas (deltaRank, deltaFile), ordre NORMATIF aligné Java
# `MoveEncoding.KNIGHT_DELTAS`.
KNIGHT_DELTAS: tuple[tuple[int, int], ...] = (
    (+2, +1),
    (+1, +2),
    (-1, +2),
    (-2, +1),
    (-2, -1),
    (-1, -2),
    (+1, -2),
    (+2, -1),
)

# Underpromotion file deltas P1 : capture-left, push, capture-right.
UNDERPROMO_FILE_DELTAS: tuple[int, ...] = (-1, 0, +1)

# Mapping python-chess.Move.promotion -> AlphaZero promo index (Java Move.promo convention).
# Java: 0=KNIGHT, 1=BISHOP, 2=ROOK, 3=QUEEN (cf. Move.pieceTypeToPromo).
# python-chess: KNIGHT=2, BISHOP=3, ROOK=4, QUEEN=5.
_CHESS_PROMOTION_TO_PROMO_INDEX: dict[int, int] = {
    chess.KNIGHT: 0,
    chess.BISHOP: 1,
    chess.ROOK: 2,
    chess.QUEEN: 3,
}
_PROMO_INDEX_TO_CHESS_PROMOTION: dict[int, int] = {
    v: k for k, v in _CHESS_PROMOTION_TO_PROMO_INDEX.items()
}

# Lookup tables précalculées (cohérent Java static init).
# QUEEN_DIR_LUT[(signR+1)*3 + (signF+1)] -> index direction queen (0..7) ou -1.
_QUEEN_DIR_LUT: list[int] = [-1] * 9
for _i, (_dr, _df) in enumerate(QUEEN_DELTAS):
    _QUEEN_DIR_LUT[(_dr + 1) * 3 + (_df + 1)] = _i

# KNIGHT_DELTA_LUT[(dR+2)*5 + (dF+2)] -> index knight (0..7) ou -1.
_KNIGHT_DELTA_LUT: list[int] = [-1] * 25
for _i, (_dr, _df) in enumerate(KNIGHT_DELTAS):
    _KNIGHT_DELTA_LUT[(_dr + 2) * 5 + (_df + 2)] = _i


def _signum(x: int) -> int:
    """Reproduit Integer.signum Java : -1, 0, +1."""
    return (x > 0) - (x < 0)


def _is_knight_shape(d_rank: int, d_file: int) -> bool:
    a_r = abs(d_rank)
    a_f = abs(d_file)
    return (a_r == 1 and a_f == 2) or (a_r == 2 and a_f == 1)


def encode_move(move: chess.Move, board: chess.Board) -> int:
    """Encode a chess.Move to an AlphaZero policy index in [0, 4672).

    Reproduit bit-pour-bit `MoveEncoding.encode(int move, int sideToMove)` Java.

    Args:
        move: chess.Move à encoder. Promotion via move.promotion (chess.QUEEN,
              chess.ROOK, chess.BISHOP, chess.KNIGHT ou None).
        board: chess.Board pour déterminer le side-to-move (mirror XOR 56 si BLACK).

    Returns:
        Index in [0, 4672).

    Raises:
        ValueError: si delta du coup ne correspond à aucun pattern (queen,
                    knight, underpromo) — coup invalide pour AlphaZero encoding.
    """
    from_sq = move.from_square
    to_sq = move.to_square
    promotion = move.promotion  # None or chess.PAWN/KNIGHT/BISHOP/ROOK/QUEEN

    # Mirror XOR 56 si BLACK to move (perspective P1).
    if board.turn == chess.BLACK:
        from_sq ^= 56
        to_sq ^= 56

    # chess.SQUARES indexe LSB=a1=0 ; rank = sq // 8, file = sq % 8.
    d_rank = (to_sq // 8) - (from_sq // 8)
    d_file = (to_sq % 8) - (from_sq % 8)

    plane_index: int

    if promotion is not None and promotion != chess.QUEEN:
        # Underpromotion (plans 64..72). promo : 0=N, 1=B, 2=R.
        direction_index = d_file + 1  # capture-left=0, push=1, capture-right=2
        if direction_index < 0 or direction_index > 2:
            raise ValueError(f"Underpromotion deltaFile invalide : {d_file} (attendu : -1, 0, +1)")
        promo_index = _CHESS_PROMOTION_TO_PROMO_INDEX[promotion]
        plane_index = 64 + direction_index * 3 + promo_index
    elif _is_knight_shape(d_rank, d_file):
        knight_idx = _KNIGHT_DELTA_LUT[(d_rank + 2) * 5 + (d_file + 2)]
        plane_index = 56 + knight_idx
    else:
        # Queen-style (inclut promotion en dame, EP, castling, normal slider/king/pawn-push).
        sign_r = _signum(d_rank)
        sign_f = _signum(d_file)
        direction = _QUEEN_DIR_LUT[(sign_r + 1) * 3 + (sign_f + 1)]
        if direction < 0:
            raise ValueError(
                f"Coup inencodable : delta queen-style invalide (d_rank={d_rank}, d_file={d_file})"
            )
        distance = max(abs(d_rank), abs(d_file))
        if distance < 1 or distance > 7:
            raise ValueError(f"Distance queen-style hors plage : {distance} (attendu : 1..7)")
        plane_index = direction * 7 + (distance - 1)

    return from_sq * POLICY_PLANES + plane_index


def decode_move(index: int, board: chess.Board) -> chess.Move:  # noqa: PLR0912
    """Decode an AlphaZero policy index back to a chess.Move in the given position.

    Reproduit `MoveEncoding.decode(int policyIndex, Position position)` Java.

    Args:
        index: in [0, 4672).
        board: chess.Board pour résoudre le type de coup (NORMAL/EN_PASSANT/
               CASTLING/PROMOTION en dame).

    Returns:
        chess.Move correspondant. Légalité dans la position N'EST PAS vérifiée
        (cohérent contrat Java : decode produit un Move syntaxiquement valide,
        la légalité est responsabilité du caller via board.is_legal).

    Raises:
        ValueError: si index hors [0, 4672).
    """
    if index < 0 or index >= POLICY_INDICES:
        raise ValueError(f"index hors plage : {index} (attendu : 0..{POLICY_INDICES - 1})")

    from_p1 = index // POLICY_PLANES
    plane = index % POLICY_PLANES

    is_underpromo = plane >= 64
    is_knight = (not is_underpromo) and plane >= 56

    promo_index: int | None = None

    if is_underpromo:
        direction_index = (plane - 64) // 3
        promo_index = (plane - 64) % 3  # 0=N, 1=B, 2=R
        d_file_p1 = UNDERPROMO_FILE_DELTAS[direction_index]
        d_rank_p1 = +1  # P1 avance toujours vers le haut en perspective normalisée
    elif is_knight:
        d_rank_p1, d_file_p1 = KNIGHT_DELTAS[plane - 56]
    else:
        direction = plane // 7
        distance = plane % 7 + 1
        d_r, d_f = QUEEN_DELTAS[direction]
        d_rank_p1 = d_r * distance
        d_file_p1 = d_f * distance

    to_p1 = from_p1 + d_rank_p1 * 8 + d_file_p1

    # Mirror retour si BLACK to move.
    if board.turn == chess.BLACK:
        from_sq = from_p1 ^ 56
        to_sq = to_p1 ^ 56
    else:
        from_sq = from_p1
        to_sq = to_p1

    if is_underpromo or _is_knight_shape_pre(plane):
        # Underpromotion ou knight : promotion explicite ou aucune.
        if is_underpromo:
            promotion: int | None = _PROMO_INDEX_TO_CHESS_PROMOTION[promo_index]  # type: ignore[index]
        else:
            promotion = None
    else:
        # Queen-style : déterminer type via la pièce qui bouge et la géométrie.
        piece = board.piece_at(from_sq)
        if piece is None:
            # Pas de pièce sur from_sq : coup syntaxiquement non interprétable
            # (probablement un index décodant un coup illégal dans la position).
            # On retourne quand même un Move "brut" sans promotion pour cohérence
            # avec le contrat Java (decode produit un Move ; caller vérifie légalité).
            promotion = None
        elif piece.piece_type == chess.PAWN and chess.square_rank(to_sq) in (0, 7):
            promotion = chess.QUEEN
        else:
            promotion = None

    return chess.Move(from_sq, to_sq, promotion=promotion)


def _is_knight_shape_pre(plane: int) -> bool:
    """Helper : un plane in [56, 64) est knight-shape."""
    return 56 <= plane < 64


def visits_to_policy_target(
    visits: dict[str, int],
    board: chess.Board,
) -> npt.NDArray[np.float32]:
    """Convert {uci_move: visit_count} to normalized policy target (4672,) float32.

    Args:
        visits: dict UCI move -> visit count (extrait `info string visits` UCI v1.2.0+).
        board: chess.Board pour encoder chaque move dans la position.

    Returns:
        np.ndarray (4672,) float32, sum ≈ 1.0 (training) OU sum == 0.0 (terminal /
        visits vide).
    """
    target = np.zeros(POLICY_INDICES, dtype=np.float32)
    if not visits:
        return target  # Terminal ou simulationsCount=0
    total = sum(visits.values())
    if total == 0:
        return target
    for uci_move, count in visits.items():
        move = chess.Move.from_uci(uci_move)
        idx = encode_move(move, board)
        target[idx] = count / total
    return target

File: nanozero-nn/src/main/java/org/nanozero/nn/MoveEncoding.java
package org.nanozero.nn;

import org.nanozero.board.Color;
import org.nanozero.board.Move;
import org.nanozero.board.MoveType;
import org.nanozero.board.Piece;
import org.nanozero.board.PieceType;
import org.nanozero.board.Position;
import org.nanozero.board.Square;

/**
 * Mapping bidirectionnel {@code Move ↔ index policy AlphaZero (4672 indices)} (cf. SPEC §3.5,
 * §4.2.5, §5.5).
 *
 * <p>Le format est figé : 73 plans par case d'origine × 64 cases = 4672 indices.
 *
 * <ul>
 *   <li>plans 0..55 : queen-style (8 directions × 7 distances)
 *   <li>plans 56..63 : knight (8 deltas)
 *   <li>plans 64..72 : underpromotions (3 directions × 3 sous-pièces N/B/R)
 * </ul>
 *
 * <p>L'inversion de perspective (XOR 56 sur les cases) est appliquée si {@code sideToMove == BLACK}
 * pour ramener le coup en perspective P1 (le réseau prédit toujours du point de vue de P1, le côté
 * au trait courant ; cf. SPEC §3.5 et §7.3 SPEC-board).
 *
 * <p>Invariants :
 *
 * <ul>
 *   <li>{@code I-ME-1} : pour tout coup légal {@code m} dans une position {@code p} avec côté au
 *       trait {@code s = p.sideToMove()}, {@code decode(encode(m, s), p) == m}.
 *   <li>{@code I-ME-2} : {@code decodePolicy} produit une distribution dont la somme des {@code n}
 *       premières entrées est exactement {@code 1.0} (modulo erreur flottante &lt; 1e-6).
 * </ul>
 *
 * <p>Classe non instanciable, conforme aux conventions du module.
 */
public final class MoveEncoding {

  /** Nombre total d'indices de policy (4672). */
  public static final int POLICY_INDICES = 4672;

  /** Nombre de plans par case d'origine (73). */
  public static final int POLICY_PLANES = 73;

  private MoveEncoding() {
    throw new AssertionError("Non-instantiable");
  }

  // -----------------------------------------------------------------------------------------
  // Tables figées (cf. SPEC §5.5.2)
  // -----------------------------------------------------------------------------------------

  /**
   * Directions queen-style en {@code (deltaRank, deltaFile)} : N, NE, E, SE, S, SW, W, NW. Ordre
   * normatif §5.5.2.
   */
  static final int[][] QUEEN_DELTAS = {
    {+1, 0}, {+1, +1}, {0, +1}, {-1, +1},
    {-1, 0}, {-1, -1}, {0, -1}, {+1, -1}
  };

  /** Deltas du cavalier en {@code (deltaRank, deltaFile)}, ordre normatif §5.5.2. */
  static final int[][] KNIGHT_DELTAS = {
    {+2, +1}, {+1, +2}, {-1, +2}, {-2, +1},
    {-2, -1}, {-1, -2}, {+1, -2}, {+2, -1}
  };

  /**
   * Deltas de fichier pour underpromotion : 3 directions (capture-left, push, capture-right) du
   * point de vue P1.
   */
  static final int[] UNDERPROMO_FILE_DELTAS = {-1, 0, +1};

  /** Pièces de sous-promotion (KNIGHT, BISHOP, ROOK), index 0..2. */
  static final int[] UNDERPROMO_PIECES = {PieceType.KNIGHT, PieceType.BISHOP, PieceType.ROOK};

  // Lookup tables précalculées au static init pour éviter les recherches linéaires en hot path.

  /**
   * {@code QUEEN_DIR_LUT[(signRank+1)*3 + (signFile+1)]} retourne l'index queen-direction (0..7) ou
   * {@code -1} si la combinaison de signes est invalide ({@code (0,0)}).
   */
  private static final int[] QUEEN_DIR_LUT = new int[9];

  /**
   * {@code KNIGHT_DELTA_LUT[(dR+2)*5 + (dF+2)]} retourne l'index knight-delta (0..7) ou {@code -1}
   * si {@code (dR, dF)} n'est pas un saut de cavalier.
   */
  private static final int[] KNIGHT_DELTA_LUT = new int[25];

  static {
    java.util.Arrays.fill(QUEEN_DIR_LUT, -1);
    for (int i = 0; i < QUEEN_DELTAS.length; i++) {
      int dR = QUEEN_DELTAS[i][0];
      int dF = QUEEN_DELTAS[i][1];
      QUEEN_DIR_LUT[(dR + 1) * 3 + (dF + 1)] = i;
    }
    java.util.Arrays.fill(KNIGHT_DELTA_LUT, -1);
    for (int i = 0; i < KNIGHT_DELTAS.length; i++) {
      int dR = KNIGHT_DELTAS[i][0];
      int dF = KNIGHT_DELTAS[i][1];
      KNIGHT_DELTA_LUT[(dR + 2) * 5 + (dF + 2)] = i;
    }
  }

  // -----------------------------------------------------------------------------------------
  // Encode
  // -----------------------------------------------------------------------------------------

  /**
   * Encode un coup en index policy AlphaZero. L'indice dépend du from-square et du type de coup
   * (queen-style, knight-style, ou underpromotion). L'inversion de perspective ({@code XOR 56}) est
   * appliquée si {@code sideToMove == BLACK} (cf. SPEC §5.5.3).
   *
   * @param move coup encodé selon le format {@link Move} 16-bit
   * @param sideToMove côté au trait courant ({@link Color#WHITE} ou {@link Color#BLACK})
   * @return index dans {@code [0..4671]}
   * @throws IllegalArgumentException si le coup est inencodable (delta ne correspondant à aucun
   *     type)
   */
  public static int encode(int move, int sideToMove) {
    int from = Move.from(move);
    int to = Move.to(move);
    int type = Move.type(move);
    int promo = Move.promo(move);

    if (sideToMove == Color.BLACK) {
      from ^= 56;
      to ^= 56;
    }

    int dRank = Square.rank(to) - Square.rank(from);
    int dFile = Square.file(to) - Square.file(from);

    int planeIndex;
    if (type == MoveType.PROMOTION && promo != 3) {
      // Underpromotion (planes 64-72). promo : 0=N, 1=B, 2=R.
      int directionIndex = dFile + 1;
      if (directionIndex < 0 || directionIndex > 2) {
        throw new IllegalArgumentException(
            "Underpromotion deltaFile invalide : " + dFile + " (attendu : -1, 0, +1)");
      }
      planeIndex = 64 + directionIndex * 3 + promo;
    } else if (isKnightShape(dRank, dFile)) {
      int knightIdx = KNIGHT_DELTA_LUT[(dRank + 2) * 5 + (dFile + 2)];
      planeIndex = 56 + knightIdx;
    } else {
      // Queen-style (inclut promotion en dame, EP, castling, normal slider/king/pawn-push).
      int signR = Integer.signum(dRank);
      int signF = Integer.signum(dFile);
      int direction = QUEEN_DIR_LUT[(signR + 1) * 3 + (signF + 1)];
      if (direction < 0) {
        throw new IllegalArgumentException(
            "Coup inencodable : delta queen-style invalide (dRank="
                + dRank
                + ", dFile="
                + dFile
                + ")");
      }
      int distance = Math.max(Math.abs(dRank), Math.abs(dFile));
      if (distance < 1 || distance > 7) {
        throw new IllegalArgumentException(
            "Distance queen-style hors plage : " + distance + " (attendu : 1..7)");
      }
      planeIndex = direction * 7 + (distance - 1);
    }

    return from * POLICY_PLANES + planeIndex;
  }

  // -----------------------------------------------------------------------------------------
  // Decode
  // -----------------------------------------------------------------------------------------

  /**
   * Décode un index policy en coup, dans le contexte d'une position donnée. La position est
   * nécessaire pour distinguer les types de coup (NORMAL, EN_PASSANT, CASTLING, PROMOTION en dame).
   * Conforme §5.5.4.
   *
   * @param policyIndex index dans {@code [0..4671]}
   * @param position position courante pour résoudre le type de coup
   * @return coup encodé selon le format {@link Move} 16-bit
   * @throws IllegalArgumentException si {@code policyIndex} est hors plage
   */
  public static int decode(int policyIndex, Position position) {
    if (policyIndex < 0 || policyIndex >= POLICY_INDICES) {
      throw new IllegalArgumentException(
          "policyIndex hors plage : "
              + policyIndex
              + " (attendu : 0.."
              + (POLICY_INDICES - 1)
              + ")");
    }
    int sideToMove = position.sideToMove();
    int fromP1 = policyIndex / POLICY_PLANES;
    int plane = policyIndex % POLICY_PLANES;

    int dRankP1;
    int dFileP1;
    int type;
    int promo;
    boolean isUnderpromo = plane >= 64;
    boolean isKnight = !isUnderpromo && plane >= 56;

    if (isUnderpromo) {
      int directionIndex = (plane - 64) / 3;
      int pieceIndex = (plane - 64) % 3;
      dFileP1 = UNDERPROMO_FILE_DELTAS[directionIndex];
      // P1 (côté au trait) avance toujours vers le haut en perspective normalisée.
      dRankP1 = +1;
      type = MoveType.PROMOTION;
      promo = pieceIndex; // 0=N, 1=B, 2=R
    } else if (isKnight) {
      int[] delta = KNIGHT_DELTAS[plane - 56];
      dRankP1 = delta[0];
      dFileP1 = delta[1];
      type = MoveType.NORMAL;
      promo = 0;
    } else {
      int direction = plane / 7;
      int distance = plane % 7 + 1;
      int[] dir = QUEEN_DELTAS[direction];
      dRankP1 = dir[0] * distance;
      dFileP1 = dir[1] * distance;
      // type/promo déterminés ci-dessous selon la position
      type = -1;
      promo = 0;
    }

    int toP1 = fromP1 + dRankP1 * 8 + dFileP1;
    int from;
    int to;
    if (sideToMove == Color.BLACK) {
      from = fromP1 ^ 56;
      to = toP1 ^ 56;
    } else {
      from = fromP1;
      to = toP1;
    }

    if (!isUnderpromo && !isKnight) {
      // Queen-style : type déterminé par la pièce qui bouge et la géométrie du coup.
      long fromBB = 1L << from;
      int movingPieceType = pieceTypeAt(position, fromBB, sideToMove);
      int absFileDiff = Math.abs(Square.file(to) - Square.file(from));

      if (movingPieceType == PieceType.KING && absFileDiff == 2) {
        type = MoveType.CASTLING;
      } else if (movingPieceType == PieceType.PAWN && to == position.epSquare()) {
        type = MoveType.EN_PASSANT;
      } else if (movingPieceType == PieceType.PAWN
          && (Square.rank(to) == 0 || Square.rank(to) == 7)) {
        type = MoveType.PROMOTION;
        promo = 3; // queen
      } else {
        type = MoveType.NORMAL;
        promo = 0;
      }
    }

    return Move.encode(from, to, type, promo);
  }

  /**
   * Retourne le type de pièce ({@link PieceType#PAWN}..{@link PieceType#KING}) du côté donné qui
   * occupe la case bit-encodée par {@code sqBB}, ou {@link PieceType#NONE} si vide.
   */
  private static int pieceTypeAt(Position pos, long sqBB, int color) {
    for (int pt = 0; pt < PieceType.NB_PIECE_TYPES; pt++) {
      if ((pos.pieceBB(Piece.make(color, pt)) & sqBB) != 0L) {
        return pt;
      }
    }
    return PieceType.NONE;
  }

  /** Indique si {@code (dRank, dFile)} est un saut de cavalier (L-shape 1×2 ou 2×1). */
  private static boolean isKnightShape(int dRank, int dFile) {
    int aR = Math.abs(dRank);
    int aF = Math.abs(dFile);
    return (aR == 1 && aF == 2) || (aR == 2 && aF == 1);
  }

  // -----------------------------------------------------------------------------------------
  // decodePolicy : softmax masqué sur coups légaux
  // -----------------------------------------------------------------------------------------

  /**
   * Applique un softmax masqué sur les coups légaux. Conforme §5.5.5.
   *
   * <p>Étapes :
   *
   * <ol>
   *   <li>Extraire les logits aux indices des coups légaux ({@code dest[i] = logits[encode(...)]})
   *   <li>Softmax stable (subtract max avant exp pour éviter overflow)
   *   <li>Normalisation à somme = 1
   * </ol>
   *
   * <p>Validation stricte des arguments avant toute logique : {@link IllegalArgumentException} avec
   * message localisé pour chaque échec. Au-delà, le contrat de {@link #encode(int, int)} prend le
   * relais sur {@code legalMoves[i]} et {@code sideToMove} (validation existant phase 1).
   *
   * <p><strong>Zéro allocation</strong> : aucun {@code new}. Le caller MCTS réutilise les buffers
   * {@code legalMoves} et {@code dest} entre nodes.
   *
   * @param logits logits bruts d'une position : {@code float[POLICY_INDICES]} (4672 entrées)
   * @param legalMoves liste de coups légaux générés via {@code MoveGen}
   * @param numLegalMoves nombre de coups légaux dans le buffer (≥ 1, ≤ {@code legalMoves.length})
   * @param sideToMove côté au trait pour l'encodage de perspective ({@code Color.WHITE} ou {@code
   *     Color.BLACK})
   * @param dest buffer destinataire : {@code float[≥ numLegalMoves]}
   * @throws IllegalArgumentException si un argument viole le contrat (null, longueurs, bornes)
   */
  public static void decodePolicy(
      float[] logits, int[] legalMoves, int numLegalMoves, int sideToMove, float[] dest) {
    if (logits == null) {
      throw new IllegalArgumentException("logits must not be null");
    }
    if (logits.length != POLICY_INDICES) {
      throw new IllegalArgumentException(
          "logits length must be " + POLICY_INDICES + ", got " + logits.length);
    }
    if (legalMoves == null) {
      throw new IllegalArgumentException("legalMoves must not be null");
    }
    if (numLegalMoves < 1) {
      throw new IllegalArgumentException(
          "decodePolicy requires at least 1 legal move (received "
              + numLegalMoves
              + "); caller must check terminal state before invoking");
    }
    if (numLegalMoves > legalMoves.length) {
      throw new IllegalArgumentException(
          "numLegalMoves ("
              + numLegalMoves
              + ") exceeds legalMoves buffer length ("
              + legalMoves.length
              + ")");
    }
    if (dest == null) {
      throw new IllegalArgumentException("dest must not be null");
    }
    if (dest.length < numLegalMoves) {
      throw new IllegalArgumentException(
          "dest length (" + dest.length + ") must be >= numLegalMoves (" + numLegalMoves + ")");
    }
    // sideToMove validé indirectement par encode() au premier coup (lève IAE si invalide).

    // 1. Extraire les logits aux indices des coups légaux.
    for (int i = 0; i < numLegalMoves; i++) {
      int idx = encode(legalMoves[i], sideToMove);
      dest[i] = logits[idx];
    }
    // 2. Softmax stable : subtract max avant exp.
    float maxLogit = Float.NEGATIVE_INFINITY;
    for (int i = 0; i < numLegalMoves; i++) {
      if (dest[i] > maxLogit) {
        maxLogit = dest[i];
      }
    }
    float sum = 0f;
    for (int i = 0; i < numLegalMoves; i++) {
      float e = (float) Math.exp(dest[i] - maxLogit);
      dest[i] = e;
      sum += e;
    }
    // 3. Normalisation.
    float invSum = 1f / sum;
    for (int i = 0; i < numLegalMoves; i++) {
      dest[i] *= invSum;
    }
  }
}
